extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/authorization/BasicRoleBasedAuthorizer.java

```diff
@@ -0,0 +1,142 @@
+{
+  private static final int DEFAULT_PERMISSION_CACHE_SIZE = 5000;
+  private final BasicAuthorizerStorageConnector dbConnector;
+  private final LoadingCache<String, Pattern> permissionPatternCache;
+  private final int permissionCacheSize;
+  private final BasicAuthDBConfig dbConfig;
+
+  @JsonCreator
+  public BasicRoleBasedAuthorizer(
+      @JacksonInject BasicAuthorizerStorageConnector dbConnector,
+      @JsonProperty("dbPrefix") String dbPrefix,
+      @JsonProperty("permissionCacheSize") Integer permissionCacheSize
+  )
+  {
+    this.dbConfig = new BasicAuthDBConfig(dbPrefix, null, null);
+    this.dbConnector = dbConnector;
+    this.permissionCacheSize = permissionCacheSize == null ? DEFAULT_PERMISSION_CACHE_SIZE : permissionCacheSize;
+    this.permissionPatternCache = Caffeine.newBuilder()
+                                          .maximumSize(this.permissionCacheSize)
+                                          .build(regexStr -> Pattern.compile(regexStr));
+  }
+
+  /**
+   * constructor for unit tests
+   */
+  public BasicRoleBasedAuthorizer(
+      BasicAuthorizerStorageConnector dbConnector,
+      String dbPrefix,
+      int permissionCacheSize
+  )
+  {
+    this.dbConnector = dbConnector;
+    this.permissionCacheSize = permissionCacheSize;
+    this.permissionPatternCache = Caffeine.newBuilder()
+                                          .maximumSize(this.permissionCacheSize)
+                                          .build(regexStr -> Pattern.compile(regexStr));
+    this.dbConfig = new BasicAuthDBConfig(dbPrefix, null, null);
+  }
+
+  @Override
+  public Access authorize(
+      AuthenticationResult authenticationResult, Resource resource, Action action
+  )
+  {
+    if (authenticationResult == null) {
+      throw new IAE("WTF? authenticationResult should never be null.");
+    }
+
+    List<Map<String, Object>> permissions = dbConnector.getPermissionsForUser(dbConfig.getDbPrefix(), authenticationResult.getIdentity());
+
+    for (Map<String, Object> permission : permissions) {
+      if (permissionCheck(resource, action, permission)) {
+        return new Access(true);
+      }
+    }
+
+    return new Access(false);
+  }
+
+  public BasicAuthorizerStorageConnector getDbConnector()
+  {
+    return dbConnector;
+  }
+
+  public String getDBPrefix()
+  {
+    return dbConfig.getDbPrefix();
+  }
+
+  private boolean permissionCheck(Resource resource, Action action, Map<String, Object> permission)
+  {
+    ResourceAction permissionResourceAction = (ResourceAction) permission.get("resourceAction");
+    if (action != permissionResourceAction.getAction()) {
+      return false;
+    }
+
+    Resource permissionResource = permissionResourceAction.getResource();
+    if (permissionResource.getType() != resource.getType()) {
+      return false;
+    }
+
+    String permissionResourceName = permissionResource.getName();
+
+    Pattern resourceNamePattern = permissionPatternCache.get(permissionResourceName);
```

Comment:
The `permissionPatternCache` won't be needed if all permissions details are cached -- since the permission objects would be long-lived and could contain their own compiled Patterns.